### PR TITLE
Update PyTryFrom for PyMapping and PySequence to more accurately check types

### LINKED ISSUE #2072: PyMapping is inconsistent with typing.Mapping
### Bug Description

The [conversion table](https://pyo3.rs/v0.15.1/conversions/tables.html#argument-types) for PyMapping suggests anything that is a typing.Mapping should be able to be converted to PyMapping, this however is not the case seemly list and str are also downcast-able to a PyMapping.

This causes errors when you try to access things like items on the PyMapping as you get a AttributeError str has no attribute items.

### Steps to Reproduce

```rs
use pyo3::prelude::*;
use pyo3::types::PyMapping;

#[pyfunction]
fn downcast_(a: &PyAny) -> PyResult<&PyMapping> {
    Ok(a.downcast::<PyMapping>()?)
}

#[pymodule]
fn downcast(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(downcast_, m)?)?;

    Ok(())
}
```

```pycon
>>> import typing, downcast
>>> isinstance({}, typing.Mapping)
True
>>> downcast.downcast_({})
{}
>>> isinstance("", typing.Mapping)
False
>>> downcast.downcast_("")
''
```

### Backtrace

_No response_

### Your operating system and version

MacOS 11.1 (20C69)

### Your Python version (`python --version`)

3.9.3

### Your Rust version (`rustc --version`)

rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)

### Your PyO3 version

0.15.1

### How did you install python? Did you use a virtualenv?

venv

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by aganders3**
This is intended to fix #2072 - see further discussion in that thread.

The implementation is such that for both `PyMapping` and `PyTryFrom`, they will call into python to check `isinstance(<obj>, collections.abc.Mapping)` before successfully downcasting. Please note this is a breaking change so it is important the documentation and migration guide are clear.

Also note the implementation has changed from the initial draft of this PR. See discussion below and in #2072 for more information.
 
### TODO
- [x]  an entry in CHANGELOG.md
- [x]  docs to all new functions and / or detail in the guide
- [x]  tests for all new or changed functions

### CODE REVIEW

File: src/pyclass.rs
Comment (davidhewitt) [reply src/pyclass.rs:327]: Yeah, I thought about doing similar in the past and decided against it. I don't think it makes sense to only in select circumstances set this flag. I would prefer that we let users make this explicit, either by calling the abc `register` like we do elsewhere in this PR or hopefully use #991 to inherit from the abc.

File: src/types/mapping.rs
Comment (davidhewitt) [inline src/types/mapping.rs:135]: As a sanity check, I wonder if there is any evidence we can collect (or tests we can write) that this check is functionally equivalent to the slow path? If there's any lingering concern that they may be different, we probably need to always use the slow path.

File: tests/test_mapping.rs
Comment (davidhewitt) [reply tests/test_mapping.rs:129]: I suppose in absence of #991 users may find it useful to have this exposed as a documented API to improve discoverability?! Not completely sure, what do you think?

**Review by davidhewitt — COMMENTED** [review]
Thanks for working on this! Got a few thoughts to get us started...

File: tests/test_mapping.rs
Comment (aganders3) [reply tests/test_mapping.rs:129]: Yeah I was thinking something like `py.register_abc_subclass<Mapping>(ABC::Mapping).unwrap()` for example. Like you say it might only be useful until #991 is available but easier and more discoverable than the above.

File: src/types/mapping.rs
Comment (aganders3) [reply src/types/mapping.rs:135]: Good call - I'll give this some thought. I still think this is the right way going forward, but in the mean time (until 3.9 EOL I guess...) the behavior should be the same.

Comment (davidhewitt) [reply src/types/mapping.rs:135]: I think it's not just 3.9 EOL - we also need to be sure that we won't have behaviour differences between abi3 and non-abi3 builds (just perf improvements).

**Comment by aganders3** [comment]
Yeah it's a bit sad but it would be more consistent behavior, even if the discrepancy would be rare. Still I'll defer to your (and others') expertise and experience on this. I don't have a good feel for how big this performance hit will be and/or how common the operation is in real use.

For performance-critical situations could a user still do their own check and use `try_from_unchecked`? It might be sufficient to just document this in the `PyMapping` docs with an explanation that there is no fast perfect check for mapping types at this time.

**Comment by davidhewitt** [comment]
Sorry for the slight delay - had a backlog of things to get through.

I think consistent behaviour is most important, so I think we should go for the `isinstance` check always. As you say, users can use `unsafe` APIs to check the flags / use `try_from_unchecked` if they really need performance and are willing to accept the differences.

File: src/types/mapping.rs
Comment (davidhewitt) [inline src/types/mapping.rs:122]: Slight preference to have this return `&'py PyType` - you can do that by adding `.is_ref(py)` after the `get_or_init` call.

File: src/types/mapping.rs
Comment (davidhewitt) [inline src/types/mapping.rs:153]:
```suggestion
        if value.is_instance(get_mapping_abc(value.py()))? {
```

File: src/types/mapping.rs
Comment (davidhewitt) [inline src/types/mapping.rs:113]: Might be nice to add "this is equivalent to `collections.abc.Mapping.register(T)`,  like we have for other docs with python equivalent functionality?

File: src/types/mapping.rs
Comment (davidhewitt) [inline src/types/mapping.rs:117]:
```suggestion
        mapping_abc.call_method1(py, "register", (ty,))?;
```

**Review by davidhewitt — COMMENTED** [review]
Thanks, this is shaping up well. Posted some implementation suggestions.

I'm beginning to wonder if we should do equivalent treatment for `PySequence`? Seems like it would be good to have both for consistency.

Also I keep going back and forth on whether we should have `#[pyclass(mapping)]` automatically register with the abc. I think it's still the right choice to make users opt-in, however if they don't when upgrading to 0.17 their casts to `PyMapping` will break so we should make a very clear entry about this in the migration guide.

File: src/types/mapping.rs
Comment (aganders3) [reply src/types/mapping.rs:122]: Thanks - this is a distinction I'm still wrapping my head around a little. I'm also not-quite comfortable with lifetime specifications.

Would this make the fn signature look like this?
```
fn get_mapping_abc<'py>(py: Python<'py>) -> &'py PyType
```
which then can be replaced by placeholders:
```
fn get_mapping_abc(py: Python<'_>) -> &'_ PyType {
```

Comment (davidhewitt) [reply src/types/mapping.rs:122]: Yes that's correct. I'm don't think the placeholder is required on the reference (as there's only one possible lifetime there), so I think you can even have 

```rust
fn get_mapping_abc(py: Python<'_>) -> &PyType {
```

**Comment by davidhewitt** [comment]
Well @aganders3 nobody has raised concerns about also changing sequences, so I guess let's move ahead with that. Do you want to add to this PR, or shall I push separately?

**Comment by aganders3** [comment]
> Well @aganders3 nobody has raised concerns about also changing sequences, so I guess let's move ahead with that. Do you want to add to this PR, or shall I push separately?

I think it makes sense to add it here as there will be overlap at least in the docs/changelog for it. I'm just getting back from vacation but should be able to do that some time this week.

**Comment by davidhewitt** [comment]
That'd be great, thanks! There's still a few items to go in #2493 before we're ready to go, so there's some time to get this in. I'm crossing things off that list when I can! :)

File: guide/src/migration.md
Comment (davidhewitt) [inline guide/src/migration.md:26]: I wonder if they should just be named `PySequence::register` and `PyMapping::register`?

File: src/types/sequence.rs
Comment (davidhewitt) [inline src/types/sequence.rs:309]: I wonder if it would be nice to avoid the panics by storing `PyResult<Py<PyType>>` in the static? Either that or we could finish off #2398 .

File: src/types/sequence.rs
Comment (davidhewitt) [inline src/types/sequence.rs:327]: Is `unwrap_or(false)` correct here? I would have thought it would be better to just return the error.

**Review by davidhewitt — COMMENTED** [review]
Thanks, this looks really solid. Just a couple of final things to decide on before we merge this in. I've commented on the sequence implementation, the equivalent ideas apply to the mapping implementation.

File: guide/src/migration.md
Comment (aganders3) [reply guide/src/migration.md:26]: Yeah I like that - it matches the Python function.

File: src/types/sequence.rs
Comment (aganders3) [reply src/types/sequence.rs:327]: This would make sense but I wasn't sure how. Do I need to implement `From<err::PyErr>` for `PyDowncastError`?

File: src/types/sequence.rs
Comment (aganders3) [reply src/types/sequence.rs:309]: This is a good idea. The public API (`register`) already returns a Result so this should be pretty straightforward I think.

File: src/types/sequence.rs
Comment (davidhewitt) [reply src/types/sequence.rs:327]: Oh I see, we can't express that with the current return type. Maybe just leave a TODO in the code and we can figure that out another time?

Comment (aganders3) [reply src/types/sequence.rs:327]: _That_ I know how to implement haha.

**Comment by aganders3** [comment]
Thanks again for the review. It was trickier than I thought to change the static to hold a `PyResult`, but I'm getting a better feel for result and error types now and hopefully what I did makes sense. Does the `clone_ref` negate the benefit of the `GILOnceCell`? Without that the return was `Result<&PyType, &PyErr>` which was more awkward to use.

I rebased on `main` again so hopefully it's still green. Please take another look when you have a chance!

File: src/types/mapping.rs
Comment (davidhewitt) [inline src/types/mapping.rs:133]: The overall approach is definitely correct, you can actually do this bit in one line:

```suggestion
        .map_or_else(|e| e.clone_ref(py), |t| t.as_ref(py))
```

File: guide/src/migration.md
Comment (davidhewitt) [inline guide/src/migration.md:10]: Looks the the wording in the first line might have meant to be `PyMapping` and `PySequence`? I suggest rephrasing like this:

```suggestion
### Type checks have been changed for `PyMapping` and `PySequence` types

Previously the type checks for `PyMapping` and `PySequence` (implemented in `PyTryFrom`)
```

**Review by davidhewitt — COMMENTED** [review]
Looking brilliant! Just two final tiny suggestions and then let's merge this. Thanks again for a really solid PR here 👍

File: src/types/mapping.rs
Comment (aganders3) [reply src/types/mapping.rs]: I had to add `Ok` and `Err` because `map_or_else` takes a `Result<T, E> -> U`, but otherwise this worked - thanks!

**Comment by aganders3** [comment]
> Thanks, this looks really solid. Just a couple of final things to decide on before we merge this in. I've commented on the sequence implementation, the equivalent ideas apply to the mapping implementation.

Thanks for all the feedback and review!

**Review by davidhewitt — APPROVED** [review]
Looks great, thanks very much again!

**Comment by Gobot1234** [comment]
@aganders3 thank you so much for fixing this!

**Pull request merged**
